url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+    }
+    @SuppressWarnings("unchecked")
+	private JSONObject getPingJson(Contact target) {
+		JSONObject message = new JSONObject();
+	  	message.put("type", "PING");
+	  	message.put("sender_id", String.valueOf(nodeID));
+	  	message.put("target_id",String.valueOf(target.getId()));
+	  	return message;
+	}
+    
+    
+    private Contact getClosestNode(int id){// that is not current
+    	Contact closest = null;
+    	int smallest_diff = Integer.MAX_VALUE;
+    	for (Map.Entry<Integer, Contact> entry : routingTable.entrySet()){                                                                                                                                                                            /**/
+    		int curr_diff = Math.abs(entry.getKey() - id);
+    		if(curr_diff < smallest_diff){
+    			smallest_diff = curr_diff;
+    			closest = entry.getValue();
+    		}
+		}
+    	return closest;
+    }
+	private void singleUpdateRouteTable(int sender_id, String hostAddress) {
+        if(sender_id == nodeID) return;// current should not be in the table
+		long currTime = System.currentTimeMillis();
+		if(routingTable.containsKey(sender_id)){
+			routingTable.get(sender_id).setTimestamp(currTime);
+			routingTable.get(sender_id).setIp(hostAddress);
+		}else{
+			routingTable.put(sender_id, new Contact(sender_id, hostAddress));
+		}	
+	}
+	private void singleUpdateRouteTable(int sender_id) {
+        if(sender_id == nodeID) return;
+		long currTime = System.currentTimeMillis();
+		if(routingTable.containsKey(sender_id)){
+			routingTable.get(sender_id).setTimestamp(currTime);
+		}else{
+			routingTable.put(sender_id, new Contact(sender_id, "127.0.0.1"));
+		}
+	}
+    private void updateRouteTable(JSONArray table){
+    	for(Object row: table){
+    		JSONObject tableNode = (JSONObject) row;
+    		int nodeID = Integer.valueOf((String)tableNode.get("node_id"));
+    		Contact tableContact = new Contact(nodeID, (String)tableNode.get("ip_addresss"));
+    		tableContact.setTimestamp(System.currentTimeMillis());
+    		routingTable.put(nodeID,tableContact);
+    	}
+    }
+    /*
+    * Put an internal periodic update routing table message for the controller to handle.
+    * Delegated to the Controller rather than here in the run method to prevent mutual exclusion
+    * problems on the routing table.Otherwise the table has to be locked if there is more than
+    * one point of entry to it.
+    * UPDATE: the rule no longer applies as the routingTable was changed 
+    * to ConcurrentHashMap so different threads do not have to worry about the table consistency.
+    * */
+    private class RefreshRoutingTable extends TimerTask {
+        @Override
+        public void run() {
+            try {
+                JSONObject message =  new JSONObject();
+                message.put("type","REFRESH_ROUTING_TABLE");
+                receiveQueue.put(jsonToDatagram(message,new InetSocketAddress("",0)));
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)